Add scalar-schema tool (#81)
scalar-labs/scalardb: tools/scalar-schema/README.md

# Schema Tool for Scalar DB

> This tool makes schemas on Cassandra or Cosmos DB for [Scalar DB](https://github.com/scalar-labs/scalardb).

Reviewer: ```suggestion
This tool creates Scalar DB schemas on Cosmos DB and Cassandra.
```
I changed the order because the following doc always explains Cosmos DB first.

Author: Thank you! I fixed it.

>   - This tool creates databases(collections) and tables(containers), also inserts metadata which is required by Scalar DB for Cosmos DB

Reviewer: Can you have a corresponding description for Cassandra?

Author: Thank you! I added a description.

# Usage

## Build & Run
### Build a standalone jar
```console
$ cd /path/to/scalardb/tools/scalar-schema
$ lein uberjar
```

### Create tables
```console
# For Cosmos DB
$ java -jar target/scalar-schema.jar --cosmos -h <YOUR_ACCOUNT_URI> -p <YOUR_ACCOUNT_PASSWORD> -f schema.json [-r BASE_RESOURCE_UNIT]
```
  - `-r BASE_RESOURCE_UNIT` is an option. You can specify the RU of each database. The maximum RU in tables in the database will be set. If you don't specify RU of tables, the database RU will be set with this option. When you use transaction function, the RU of the coordinator table of Scalar DB is specified by this option. By default, it's 400.

```console
# For Cassandra

> $ java -jar target/scalar-schema.jar --cassandra -h <CASSANDRA_IP> -u <CASSNDRA_USER> -p <CASSANDRA_PASSWORD> -f schema.json [--network-strategy <NETWORK_STRATEGY> --replication-factor <REPLICATION_FACTOR>]

Reviewer: Just a question. `--network-strategy` needs to start with double dashes?  Can it also be `-ns` or `-rf` if they are configured so?

Author: `clojure/tools.cli` doesn't support multi-characters with a single dash. I added short options for them.

```
% java -jar target/scalar-schema.jar --help
      --cassandra                                   Operate for Cassandra
      --cosmos                                      Operate for Cosmos DB
  -h, --host DB_HOST                                Address of Cassandra like IP or URI address of your Cosmos DB account
  -u, --user USERNAME                               Username of the database. This option is ignored when Cosmos DB.
  -p, --password ACCOUNT_PASSWORD                   Password of Cassandra or Cosmos DB account
  -f, --schema-file SCHEMA_JSON                     Schema file
  -r, --ru RESOURCE_UNIT                       400  Base RU for each table on Cosmos DB. The RU of the coordinator for Scalar DB transaction is specified by this option. This option is ignored when Cassandra.
  -R, --replication-factor REPLICATION_FACTOR  1    The number of replicas. This options is ignored when Cosmos DB.
  -n, --network-strategy NETWORK_STRATEGY           The network topology strategy. SimpleStrategy or NetworkTopologyStrategy. This options is ignored when Cosmos DB.
  -D, --delete-all                                  All database will be deleted, if this is enabled.
      --help
```

```

### Delete all tables
```console
# For Cosmos DB
$ java -jar target/scalar-schema.jar --cosmos -h <ACCOUNT_URI> -p <KEY> -D

# For Cassandra
$ java -jar target/scalar-schema.jar --cassandra -h <CASSANDRA_IP> -u <CASSNDRA_USER> -p <CASSANDRA_PASSWORD> -D
```

### Show help
```console
$ java -jar target/scalar-schema.jar --help
```

### Sample schema file

> ```json

Reviewer: I feel it's less confusing to separate the sample into Cassandra sample and Cosmos DB sample. Thought?

Author: You have a good point.
This sample shows the same JSON file can be used.
To avoid confusion, I added comments.

# Schema Tool for Scalar DB
This tool creates Scalar DB schemas on Cosmos DB and Cassandra.
  - For Cosmos DB, this tool creates databases(collections) and tables(containers), also inserts metadata which is required by Scalar DB.

>   - For Cassandra, this tool create databases(keyspaces) and tables. You can specify the compaction strategy, the network topology strategy and the replication factor.

Reviewer: ```suggestion
  - For Cassandra, this tool creates databases(keyspaces) and tables. You can specify the compaction strategy, the network topology strategy, and the replication factor as well.
```

```
  - `<NETWORK_STRATEGY>` should be `SimpleStrategy` or `NetworkTopologyStrategy`

### Delete all tables
```console
# For Cosmos DB
$ java -jar target/scalar-schema.jar --cosmos -h <YOUR_ACCOUNT_URI> -p <YOUR_ACCOUNT_PASSWORD> -D
```

```console
# For Cassandra
$ java -jar target/scalar-schema.jar --cassandra -h <CASSANDRA_IP> -u <CASSNDRA_USER> -p <CASSANDRA_PASSWORD> -D
```

### Show help
```console
$ java -jar target/scalar-schema.jar --help
```

### Sample schema file
```json
{
  "sample-db.sample-table1": {
    "transaction": true,
    "partition-key": [
      "c1"
    ],
    "clustering-key": [
      "c4"
    ],
    "columns": {
      "c1": "INT",
      "c2": "TEXT",
      "c3": "INT",
      "c4": "INT",
      "c5": "BOOLEAN"
    },
    "compaction-strategy": "LCS"
  },

  "sample-db.sample-table2": {
    "transaction": false,
    "partition-key": [
      "c1"
    ],
    "clustering-key": [
      "c3",
      "c4"
    ],
    "columns": {
      "c1": "INT",
      "c2": "TEXT",
      "c3": "INT",
      "c4": "INT",
      "c5": "BOOLEAN"
    },
    "ru": 800
  }
}
```

> - `compaction-strategy` should be `STCS`, `LCS` or `TWCS`. This is ignored when Cosmos DB.

Reviewer: ```suggestion
- `compaction-strategy` should be `STCS`, `LCS` or `TWCS`. This is ignored in Cosmos DB.
```

> - This `ru` value is set for all tables on this database even if `-r BASE_RESOURCE_UNIT` is set when Cosmos DB. `ru` is ignored when Cassandra.

Reviewer: ```suggestion
- This `ru` value is set for all tables on this database even if `-r BASE_RESOURCE_UNIT` is set when Cosmos DB. `ru` is ignored in Cassandra.
```
